nixos/modules/services/audio/mpdscribble.nix

```diff
@@ -0,0 +1,236 @@
+      '';
+    };
+
+    password = mkOption {
+      default = null;
+      type = types.nullOr types.str;
+      description = ''
+        Password for the mpd daemon.
+        Attention: This will be stored in the globally readable /nix/store.
+      '';
+    };
+
+    passwordFile = mkOption {
+      default = null;
+      type = types.nullOr types.str;
+      description = ''
+        File containing the password for the mpd daemon.
+      '';
+    };
+
+    port = mkOption {
+      default = mpdCfg.network.port;
+      type = types.port;
+      description = ''
+        Port for the mpdscribble daemon to search for a mpd daemon on.
+      '';
+    };
+
+    endpoints = mkOption {
+      type = (let
+        endpoint = { name, ... }: {
+          options = {
+            url = mkOption {
+              type = types.str;
+              default = endpointUrls.${name} or "";
+              description =
+                "The url endpoint where the scrobble API is listening.";
+            };
+            username = mkOption {
+              type = types.str;
+              description = ''
+                Username for the scrobble service.
+              '';
+            };
+            password = mkOption {
+              type = types.str;
+              description = ''
+                Password, either as MD5SUM or cleartext.
+                Attention: This will be stored in the globally readable /nix/store.
+              '';
+            };
+            passwordFile = mkOption {
+              type = types.nullOr types.str;
+              description =
+                "File containing the password, either as MD5SUM or cleartext.";
+            };
+            journal = mkOption {
+              type = types.path;
+              default = "/var/lib/mpdscribble/${name}.journal";
+              description = "The path where mpdscribble will store the journal";
+            };
+          };
+        };
+      in types.attrsOf (types.submodule endpoint));
+      default = { };
+      example = {
+        "last.fm" = {
+          url = "http://post.audioscrobbler.com/";
+          username = "foo";
+          passwordFile = "/run/secrets/mpd_password";
+          journal = "/var/lib/mpdscribble/lastfm.journal";
+        };
+      };
+      description = ''
+        Endpoints to scrobble to.
+        If the endpoint is one of ${
+          concatStringsSep ", " (attrNames endpointUrls)
+        } the url is set automatically.
+      '';
+    };
+
+  };
+
+  ###### implementation
+
+  config = mkIf cfg.enable {
+    systemd.services.mpdscribble = {
+      after = [ "mpd.service" "network.target" ];
+      description = "mpdscribble mpd scrobble client";
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
+        DynamicUser = true;
+        StateDirectory = "mpdscribble";
+        RuntimeDirectory = "mpdscribble";
+        RuntimeDirectoryMode = "700";
+        ExecStartPre = preStart;
```

Comment:
This implies that the `passwordFile` options are no more secure than the `password` options you have. If a `DynamicUser` can read the file provided by the `passwordFile` option then **any** user can read this file.

You likely want to either want to use:
- [`systemd` credentials logic](https://github.com/systemd/systemd/blob/6706384a89ae0c462e7172588c80667190c4d9e2/NEWS#L320)
- some advice from [this](https://cedwards.xyz/protecting-systemd-service-secrets/) blog, or
- [prefix `ExecStartPre` with a `+`](https://www.freedesktop.org/software/systemd/man/systemd.service.html)

Reply:
Thanks! I've gone for the 3rd option for now, since it was the easiest. Maybe this should be replaced with systemd credentials logic in the future, but I think the logic in the `preStart` script is fairly minimal and safe to run as root.

Reply:
That sounds like a good idea. The credentials logic is still very new and I don't think we've used it anywhere in NixOS yet. :+1: